client-core/src/mnemonic.rs

```diff
@@ -0,0 +1,205 @@
+//! Mnemonic wrapper
+use std::fmt;
+use zeroize::Zeroize;
+
+use bip39::{Language, MnemonicType, Seed};
+use secstr::SecUtf8;
+use serde::de::{self, Deserializer, Visitor};
+use serde::{Deserialize, Serialize, Serializer};
+
+use client_common::{ErrorKind, Result, ResultExt};
+
+const MNEMONIC_TYPE: MnemonicType = MnemonicType::Words24;
+const MNEMONIC_LANGUAGE: Language = Language::English;
+
+/// Mnemonic wrapped in secures string
+pub struct Mnemonic(bip39::Mnemonic);
+
+impl Serialize for Mnemonic {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.0.phrase())
+    }
+}
+
+impl<'de> Deserialize<'de> for Mnemonic {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct StrVisitor;
+
+        impl<'de> Visitor<'de> for StrVisitor {
+            type Value = Mnemonic;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("mnemonic words as string separated by space")
+            }
+
+            #[inline]
+            fn visit_str<E>(self, value: &str) -> std::result::Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Mnemonic::from_str(value).map_err(|e| E::custom(format!("{}", e)))
+            }
+        }
+
+        deserializer.deserialize_str(StrVisitor)
```

Comment:
Seems simpler to write as ``Mnemonic::from_str(&String::deserialize(deserializer)?).map_err(...)?``.

Reply:
Thanks! I have updated the code and added test cases